File: src/lib/pitono/src/base_map.py
from .base_setup import BaseSetup
from typing import List, Any, Callable

class BaseMap(BaseSetup):
    def __init__(
        self,
        features: List[str],
        feeds: List[Any],
        validates: List[Any],
        map_cb: Callable,
        initial_values: Any,
        table_data: List[Any] = None
    ):
        # Map feeds to actions, validates to checks
        super().__init__(
            features,
            feeds,
            validates,
            map_cb,
            initial_values
        )
        self.table_data = table_data or []
    
    # Alias setup to map for TDT pattern
    async def map(
        self,
        subject: Any,
        key: str,
        test_resource_configuration,
        tester: Callable[[Any], bool],
        artifactory=None,
        suite_ndx=None
    ):
        return await super().setup(
            subject,
            key,
            test_resource_configuration,
            tester,
            artifactory,
            suite_ndx
        )
    
    # Method to get table data
    def get_table_data(self) -> List[Any]:
        return self.table_data

File: src/lib/rusto/src/base_map.rs
use crate::base_setup::BaseSetup;
use crate::types::IbddInAny;

/// BaseMap extends BaseSetup to support TDT (Table Driven Testing) pattern.
pub struct BaseMap<I: IbddInAny> {
    inner: BaseSetup<I>,
    table_data: Vec<Box<dyn std::any::Any>>,
}

impl<I: IbddInAny> BaseMap<I> {
    pub fn new(
        features: Vec<String>,
        feeds: Vec<Box<dyn std::any::Any>>,
        validates: Vec<Box<dyn std::any::Any>>,
        map_cb: I::Given,
        initial_values: Box<dyn std::any::Any>,
        table_data: Vec<Box<dyn std::any::Any>>,
    ) -> Self {
        Self {
            inner: BaseSetup::new(features, feeds, validates, map_cb, initial_values),
            table_data,
        }
    }
    
    // Alias setup to map for TDT pattern
    pub async fn map(
        &mut self,
        subject: I::Isubject,
        key: &str,
        test_resource_configuration: &crate::types::ITTestResourceConfiguration,
        tester: fn(I::Then) -> bool,
        artifactory: Option<fn(&str, &dyn std::any::Any)>,
        suite_ndx: Option<i32>,
    ) -> Option<I::Istore> {
        self.inner.setup(subject, key, test_resource_configuration, tester, artifactory, suite_ndx).await
    }
    
    // Method to get table data
    pub fn get_table_data(&self) -> &Vec<Box<dyn std::any::Any>> {
        &self.table_data
    }
    
    // Delegate methods to inner
    pub fn add_artifact(&mut self, path: String) {
        self.inner.add_artifact(path);
    }
    
    pub fn to_obj(&self) -> std::collections::HashMap<String, Box<dyn std::any::Any>> {
        self.inner.to_obj()
    }
}
